### PR TITLE
Fixed the Tiberian Sun weapon factory doors

### PR DESCRIPTION
**Pull request by Mailaender**
They will now open and show the insides properly with a new modular render overlay trait. `RenderBuildingWarFactory` was too hard-coded to the Red Alert and Tiberian Dawn sprites.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+			var bi = init.Actor.Traits.Get<BuildingInfo>();
+			var offset = FootprintUtils.CenterOffset(init.World, bi).Y + Offset;
+			yield return new SpriteActorPreview(anim, WVec.Zero, offset, p, rs.Scale);
+		}
+	}
+
+	class WithProductionDoor : INotifyBuildComplete, ITick, INotifyProduction, INotifySold, ISync, INotifyDamageStateChanged
+	{
+		WithProductionDoorInfo info;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs]: `readonly`, please.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		RenderSprites renderSprites;
+		Animation door;
+		[Sync] bool isOpen;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs]: This is a cosmetic trait, so these don't really need to be synced.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs]: Shall I also remove it from RenderBuildingWarFactory?

**Comment by pchote** [comment]
It might pay to talk to @Phrohdoh about what changes this will need for RA2 (the weapons factory has two doors with different animations), and if they aren't too difficult putting them in now.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs (hunk 2 above)
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs]: Ok, I guess so.

**Comment by Mailaender** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	class WithProductionDoorInfo : ITraitInfo, IRenderActorPreviewSpritesInfo, Requires<RenderSpritesInfo>, Requires<IBodyOrientationInfo>, Requires<BuildingInfo>
+	{
+		[Desc("Additional units relative to the center to align to the top of cell.")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs]: I can't understand this at all.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs]: Me neither. Removed it again. It shouldn't be configurable I guess.

**Comment by abcdefg30** [comment]
Code looks ok and works in TS. :+1: / :white_check_mark:

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Play an animation when a unit exits or blocks the exit after production finished.")]
+	class WithProductionDoorInfo : ITraitInfo, IRenderActorPreviewSpritesInfo, Requires<RenderSpritesInfo>, Requires<IBodyOrientationInfo>, Requires<BuildingInfo>
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs]: The convention we have set is to call these traits `With*Overlay`.  You could rename this to `WithProductionDoorOverlay`, or merge its behaviour in to the existing `WithProductionOverlay` trait.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	{
+		public readonly string IdleSequence = "idle-door";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoor.cs]: This should be just `Sequence`.

**Comment by Mailaender** [comment]
Rebased and renamed.

**Comment by reaperrr** [comment]
Needs a rebase.

One question: Since (most likely) none of the currently shipping mods will use this and the RA2 mod will need ```OpenRA.Mods.TS``` anyway (because of stuff like voxels, and probably a couple of other things), wouldn't it make more sense to move this to the TS dll and namespace? On a sidenote, that should make this PR TS-only and no longer fall under the feature-freeze.

**Comment by Mailaender** [comment]
Rebased and moved to `OpenRA.Mods.TS`. I can't guarantee this will work with RA2, but I worked towards that it should also work with two door traits per building.

**Comment by pchote** [comment]
@Mailaender suggested in IRC that this could be used to replace RenderBuildingWarFactory completely.  I agree, and would like to see this done before merging.

**Comment by Mailaender** [comment]
Updated. This also fixes the Dune 2000 refinery being rendered twice. A regression from https://github.com/OpenRA/OpenRA/pull/7477. I always wondered why the shadow is so strong.

**Comment by pchote** [comment]
Could we please get an upgrade rule that replaces `-RenderBuilding:` and `RenderBuildingWarFactory:` with `WithProductionDoorOverlay:`?  This will likely break custom maps otherwise.

**Comment by Mailaender** [comment]
That is pretty tricky as that is not always the right update (see the fake buildings and refineries).

**Comment by Mailaender** [comment]
Added an upgrade rule and ran it (did not forget any custom map in the repository).

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+			if (door.CurrentSequence != null)
+				door.ReplaceAnim(RenderSprites.NormalizeSequence(door, e.DamageState, door.CurrentSequence.Name));
+		}
+
+		public void UnitProduced(Actor self, Actor other, CPos exit)
+		{
+			door.PlayThen(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.BuildSequence), () => { isOpen = true; openExit = exit; });
+		}
+
+		public void Selling(Actor self) { renderSprites.Remove("door_overlay_{0}".F(info.Sequence)); }
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: Convention is to make the overlay invisible (by setting a `buildComplete` bool to false) instead of explicitly removing the animation.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+			door = new Animation(init.World, renderSprites.GetImage(init.Self));
+			var buildingInfo = init.Self.Info.Traits.Get<BuildingInfo>();
+
+			var offset = FootprintUtils.CenterOffset(init.World, buildingInfo).Y + 512;
+			renderSprites.Add("door_overlay_{0}".F(info.Sequence), new AnimationWithOffset(door, null, () => !buildComplete, offset));
+		}
+
+		public void BuildingComplete(Actor self)
+		{
+			door.Play(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.Sequence));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: Convention is to do this in the constructor instead of here.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+		public void DamageStateChanged(Actor self, AttackInfo e)
+		{
+			if (door.CurrentSequence != null)
+				door.ReplaceAnim(RenderSprites.NormalizeSequence(door, e.DamageState, door.CurrentSequence.Name));
+		}
+
+		public void UnitProduced(Actor self, Actor other, CPos exit)
+		{
+			door.PlayThen(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.BuildSequence), () => { isOpen = true; openExit = exit; });
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: This should only be played if `!isOpen`.

**Comment by pchote** [comment]
This adds some nice polish and code cleanups.  Still a couple more minor code issues to address before merging, though.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs (hunk 8 above)
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: It probably also wants to set `isOpen = true` at the start of the animation to avoid glitches.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+		{
+			door.Play(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.Sequence));
+			buildComplete = true;
+		}
+
+		public void Tick(Actor self)
+		{
+			if (isOpen && !self.World.ActorMap.GetUnitsAt(openExit).Any(a => a != self))
+			{
+				isOpen = false;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: and then this should be set after the reverse animation plays.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: That won't work. It will then only close when a new actor is produced which is glitchy in a different kind of way and I can't get that fixed so I will keep the legacy behavior for now.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: Ok. You can solve the glitches in a different way by adding bools for `shouldClose = isOpen && self.World.ActorMap.GetUnitsAt(openExit).Any(a => a != self)` and `animating` = (set true at the start of an animation and false at the end).

The code here can then check `if (!animating && shouldClose)` and the check in `UnitProduced` becomes `if (!animating && !isOpen)`.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: Or perhaps simpler, maintain separate `isOpen` and `shouldBeOpen` bools, and then handle the transition in both directions from Tick.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: This is fiddly to get right. Tried all your suggestions, but it still doesn't work quite right especially with fast build cheats on. Let's do this in another pull request.

**Comment by pchote** [comment]
After some tweaking I came up with 04f8dda26a002b80dd7dca90e8a9e4468f931d8b as a simpler implementation that sidesteps the possibility of animation glitches.

**Comment by Mailaender** [comment]
Integrated your commit here.

**Comment by Mailaender** [comment]
I may be able to use your new animation method to complete https://github.com/OpenRA/OpenRA/pull/7998.

**Comment by pchote** [comment]
:+1: from me, but my patch significantly changes this so i'll wait for somebody else to give a final review and merge.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+		public object Create(ActorInitializer init) { return new WithProductionDoorOverlay(init.Self, this); }
+
+		public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer init, RenderSpritesInfo rs, string image, int facings, PaletteReference p)
+		{
+			var anim = new Animation(init.World, image, () => 0);
+			anim.PlayFetchIndex(Sequence, () => 0);
+
+			var bi = init.Actor.Traits.Get<BuildingInfo>();
+			var offset = FootprintUtils.CenterOffset(init.World, bi).Y + 512; // Additional 512 units move from center -> top of cell
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: I really don't like the magic number here. The comment doesn't help me understand what is happening either.
Can we make this configurable (with a nice description maybe?), or at least add a better comment what is happening here and why we have a hadrcoded number.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		public WithProductionDoorOverlay(Actor self, WithProductionDoorOverlayInfo info)
+		{
+			var renderSprites = self.Trait<RenderSprites>();
+			door = new Animation(self.World, renderSprites.GetImage(self));
+			door.PlayFetchDirection(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.Sequence),
+				() => desiredFrame - door.CurrentFrame);
+
+			var buildingInfo = self.Info.Traits.Get<BuildingInfo>();
+
+			var offset = FootprintUtils.CenterOffset(self.World, buildingInfo).Y + 512;
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: Same thing is happening here, I see. Still no clue what it is.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs (hunk 10 above)
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: I'm not convinced we need to deal with that right now.This is legacy code that has been moved from the original trait.

My preference (for whenever this gets dealt with) would be to remove this completely and rely on a sequence-defined z-offset.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: We had it configurable, but it only confused @Phrohdoh and it also should not be changed. Did some unsuccessful experiments already weeks ago.

File: mods/ts/sequences/structures.yaml [hunk 12]
```diff
@@ -329,7 +332,6 @@ naweap:
 	icon: nwepicon
 		Offset: 0, 0
 		UseTilesetCode: false
-# TODO: ntweap_1 & ntweap_b & ntweap_c are unused
```
Comment (penev92) [inline mods/ts/sequences/structures.yaml]: I still don't see a `ntweap_c` anywhere.

**Comment by penev92** [comment]
Well apart from those leftover magic numbers the code looks simple enough (as I've been told, there are no spaghetti here).
Works ingame, so I'm on board.

:+1:  once that pesky `ntweap_c` is found.

**Comment by Mailaender** [comment]
It doesn't exist... 
image

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs (hunk 11 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: Okay, replaced the magic number with something else.

File: OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+		public WithProductionDoorOverlay(Actor self, WithProductionDoorOverlayInfo info)
+		{
+			var renderSprites = self.Trait<RenderSprites>();
+			door = new Animation(self.World, renderSprites.GetImage(self));
+			door.PlayFetchDirection(RenderSprites.NormalizeSequence(door, self.GetDamageState(), info.Sequence),
+				() => desiredFrame - door.CurrentFrame);
+
+			var buildingInfo = self.Info.Traits.Get<BuildingInfo>();
+			var offset = FootprintUtils.CenterOffset(self.World, buildingInfo).Y + (new WRange(1).Range / 2); // move from center -> top of cell
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: This is even less readable than 512, IMO.  It also evaluates to 0, so is wrong.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Render/WithProductionDoorOverlay.cs]: Removed it again.

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/b1ffeb667c28e2be24a3d70572deb8b6fb01d297...633b487f6eb93799a6e1454bbb2949fc6428fad6)

**Comment by Mailaender** [comment]
[supplementing](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/633b487f6eb93799a6e1454bbb2949fc6428fad6...b026c715c8fcc8c012608930f17eae68b22aeb37)
